src/journal/journald-server.c

```diff
@@ -390,44 +396,34 @@ static int system_journal_open(Server *s, bool flush_requested, bool relinquish_
 
 static ManagedJournalFile* find_journal(Server *s, uid_t uid) {
         _cleanup_free_ char *p = NULL;
+        JournalStorage *storage;
         ManagedJournalFile *f;
         int r;
 
         assert(s);
 
-        /* A rotate that fails to create the new journal (ENOSPC) leaves the rotated journal as NULL.  Unless
-         * we revisit opening, even after space is made available we'll continue to return NULL indefinitely.
+        /* A rotate that fails to create the new journal (ENOSPC) leaves the rotated journal as
+         * NULL.  Unless we revisit opening, even after space is made available we'll continue to
+         * return NULL indefinitely.
          *
-         * system_journal_open() is a noop if the journals are already open, so we can just call it here to
-         * recover from failed rotates (or anything else that's left the journals as NULL).
+         * system_journal_open() is a noop if the journals are already open, so we can just call it
+         * here to recover from failed rotates (or anything else that's left the journals as NULL).
          *
          * Fixes https://github.com/systemd/systemd/issues/3968 */
         (void) system_journal_open(s, false, false);
 
-        /* We split up user logs only on /var, not on /run. If the runtime file is open, we write to it
-         * exclusively, in order to guarantee proper order as soon as we flush /run to /var and close the
-         * runtime file. */
-
-        if (s->runtime_journal)
-                return s->runtime_journal;
-
-        /* If we are not in persistent mode, then we need return NULL immediately rather than opening a
-         * persistent journal of any sort.
-         *
-         * Fixes https://github.com/systemd/systemd/issues/20390 */
-        if (!IN_SET(s->storage, STORAGE_AUTO, STORAGE_PERSISTENT))
-                return NULL;
-
         if (uid_for_system_journal(uid))
-                return s->system_journal;
+                return s->system_journal ? : s->runtime_journal;
```

Comment:
I'm not really sure to see how `find_journal()` could be called with both `s->runtime_journal` and `s->systemd_journal` not NULL. But in the new code, it can happen only when `find_journal()` is called when flushing journals to /var and in this specific case we want to write into the persistent/system journals.